### PR TITLE
nanomsg binding

### PR DESCRIPTION
**Pull request by ancaantochi**

### CODE REVIEW

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    return nn_close(socket);
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1bind
+(JNIEnv *env, jobject obj, jint socket, jstring address) {
+    const char *addr = (*env)->GetStringUTFChars(env, address, NULL);
+    
+    if (addr == NULL)
+    {
+        addr = 0;
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: Is passing a `null` string and treating that to mean as being the address `0` acceptable?

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1shutdown
+(JNIEnv *env, jobject obj, jint socket, jint endpoint) {
+    return nn_shutdown(socket, endpoint);
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1send
+(JNIEnv *env, jobject obj, jint socket, jbyteArray buffer, jint flags) {
+    jint result = -1;
+    jsize length = (*env)->GetArrayLength(env, buffer);
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: Can `buffer` be `null`? Can `GetArrayLength` fail?

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+
+JNIEXPORT jbyteArray JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1recv
+(JNIEnv *env, jobject obj, jint socket, jint flags)
+{
+    void *buf = NULL;
+    int nbytes = nn_recv(socket, &buf, NN_MSG, flags);
+    jbyteArray result = 0;
+    if (nbytes < 0) {
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: Is `nbytes` allowed to be `0`? Will calls like `NewByteArray` below be OK with `nbytes` being `0`?

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        
+        if (result == NULL) {
+            result = 0;
+        }
+        else {
+            (*env)->SetByteArrayRegion(env, result, 0, nbytes, (const jbyte*)buf);
+            jthrowable exception = (*env)->ExceptionOccurred(env);
+            if (exception) {
+                result = 0;
+                (*env)->ExceptionClear(env);
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: I am not sure, but are we sort of swallowing the exception here? Should this not get propagated back to the caller?

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                result = NULL;
+                (*env)->ExceptionClear(env);
+            }
+            else {
+                result = jmap_object;
+                jmethodID jmap_put = (*env)->GetMethodID(env, jMap_class, "put", "(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;");
+                exception = (*env)->ExceptionOccurred(env);
+                if (jmap_put == NULL || exception)
+                {
+                    result = NULL;
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c:139]: Feels weird not freeing `jmap_object` but then this is Java.  ¯\_(ツ)_/¯

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                            for (i = 0; ; ++i) {
+                                const char* name = nn_symbol(i, &value);
+                                if (name == NULL) break;
+
+                                jstring jkey = (*env)->NewStringUTF(env, name);
+                                exception = (*env)->ExceptionOccurred(env);
+                                if (jkey == NULL || exception)
+                                {
+                                    result = NULL;
+                                    (*env)->ExceptionClear(env);
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: Should we continue iterating through the loop in this case or should we bail?

Comment (avranju) [reply proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: Also in other error cases below.

File: proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+    NULL, NULL, NULL, NewStringUTF, NULL, GetStringUTFChars, NULL, GetArrayLength, NULL, NULL,
+    NULL, NULL, NewByteArray, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+    GetByteArrayElements, NULL, NULL, NULL, NULL, NULL, NULL, NULL, ReleaseByteArrayElements, NULL,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+    NULL, NULL, NULL, NULL, SetByteArrayRegion, NULL, NULL, NULL, NULL, NULL,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL
+};
+
```
Comment (avranju) [inline proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c:144]: Whoa! :eyes:

File: tools/build_java_oop.cmd [hunk 8]
```diff
@@ -13,30 +13,30 @@ set build-root=%current-path%\..
...
+REM popd
 
-pushd %build-jnano-root%\jnano
-call mvn clean install -DskipTests
-if errorlevel 1 goto :eof
-popd
+REM pushd %build-jnano-root%\jnano
+REM call mvn clean install -DskipTests
+REM if errorlevel 1 goto :eof
+REM popd
```
Comment (avranju) [inline tools/build_java_oop.cmd]: Why not just delete all the `REM` lines?

File: proxy/gateway/java/gateway-remote-module/src/main/java/com/microsoft/azure/gateway/remote/CommunicationEndpoint.java [hunk 9]
```diff
@@ -6,16 +6,14 @@
...
-
 /**
  * An endpoint that is used to communicate with the remote Gateway which uses nanomsg to send and receive messages. 
  *
  */
 class CommunicationEndpoint {
 
     private final String uri;
-    private final NanoLibrary nano;
+    private final NanomsgLibrary nano;
```
Comment (avranju) [inline proxy/gateway/java/gateway-remote-module/src/main/java/com/microsoft/azure/gateway/remote/CommunicationEndpoint.java]: Should this type sort of *commit* to the `NanomsgLibrary` type? Should the nanomsg stuff be hidden behind another interface?

File: proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c (hunk 7 above)
Comment (avranju) [reply proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c:144]: Maybe add some comments describing what this thing is?

**Review by avranju — APPROVED** [review]

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c (hunk 1 above)
Comment (ancaantochi) [reply proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: removed: result in an error in nanomsg with invalid argument in both cases

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c (hunk 2 above)
Comment (ancaantochi) [reply proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: fixed when buffer is NULL. GetArrayLength doesn't throw.

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c (hunk 3 above)
Comment (ancaantochi) [reply proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: added check for nbytes == 0

File: proxy/gateway/java/nanomsg-binding/java_nanomsg.c (hunk 6 above)
Comment (ancaantochi) [reply proxy/gateway/java/nanomsg-binding/java_nanomsg.c]: should continue iterating, but I removed setting the result to NULL

File: proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c (hunk 7 above)
Comment (ancaantochi) [reply proxy/gateway/java/nanomsg-binding/tests/java_nanomsg_ut.c:144]: done

**Pull request merged**